File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
from collections import deque

input = sys.stdin.readline


def bfs(i, j, count):
    num_apart = 1
    Q = deque()
    Q.append([i, j])
    graph[i][j] = count
    while Q:
        x, y = Q.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < N and 0 <= ny < N and graph[nx][ny] == 1:
                graph[nx][ny] = count
                Q.append([nx, ny])
                num_apart += 1
    arr_num_apart.append(num_apart)


N = int(input())
graph = []
for _ in range(N):
    graph.append(list(map(int, list(input().rstrip()))))


dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

count = 1
arr_num_apart = []
for i in range(N):
    for j in range(N):
        if graph[i][j] == 1:
            count += 1
            bfs(i, j, count)

arr_num_apart.sort()
print(count - 1)
for i in arr_num_apart:
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');

const n = +input[0];

const map = input.slice(1).map(v => v.split('').map(Number));
const visited = [...Array(n)].map(() => [...Array(n)].map(() => false));

function bfs(startX, startY) {
  let count = 1;
  const q = [[startX, startY]];
  visited[startX][startY] = true;
  const dr = [-1, 1, 0, 0];
  const dc = [0, 0, -1, 1];

  while (q.length) {
    const [r, c, num] = q.shift();
    for (let i = 0; i < 4; i++) {
      const nr = r + dr[i];
      const nc = c + dc[i];
      if (nr >= 0 && nr < n && nc >= 0 && nc < n) {
        if (map[nr][nc] === 1 && !visited[nr][nc]) {
          q.push([nr, nc]);
          count++;
          visited[nr][nc] = true;
        }
      }
    }
  }
  return count;
}

let village = 0;
const arr = [];
for (let i = 0; i < map.length; i++) {
  for (let j = 0; j < map[0].length; j++) {
    if (map[i][j] === 1 && !visited[i][j]) {
      arr.push(bfs(i, j));
      village += 1;
    }
  }
}

console.log(village);
console.log(arr.sort((a, b) => a - b).join('\n'));
